File: examples/python/authentication/sign_in_example.py
"""
Sign In Example

Demonstrates how to sign in to GeoGuessr programmatically.

⚠️  IMPORTANT WARNINGS:
1. This is for EDUCATIONAL purposes only
2. Programmatic login may violate GeoGuessr's Terms of Service
3. Never hardcode credentials in your code
4. Use proper authentication methods in production
5. Always store credentials securely

Usage:
    python sign_in_example.py

Requirements:
    pip install requests
"""

import requests
import os
import sys


def sign_in(email, password):
    """
    Sign in to GeoGuessr.

    ⚠️  WARNING: For educational purposes only

    Args:
        email (str): User email
        password (str): User password

    Returns:
        tuple: (session, user_data) or (None, None) if failed
    """
    try:
        url = 'https://www.geoguessr.com/api/v3/accounts/signin'

        session = requests.Session()
        response = session.post(
            url,
            json={
                'email': email,
                'password': password
            }
        )

        if response.status_code == 401:
            print('❌ Invalid email or password')
            return None, None

        response.raise_for_status()
        user = response.json()

        # Display success
        print(f"\n✅ Sign In Successful!\n")
        print(f"Welcome, {user['nick']}!")
        print(f"Level: {user['progress']['level']}")
        print(f"XP: {user['progress']['xp']:,}")
        print(f"Pro User: {'✅ Yes' if user['isProUser'] else '❌ No'}")

        print(f"\n🔐 Authentication Cookies Set:")
        print(f"- _ncfa - Main session cookie")
        print(f"- session - Session identifier")

        # Extract _ncfa cookie for future use
        ncfa_cookie = session.cookies.get('_ncfa')
        if ncfa_cookie:
            print(f"\n💡 Your _ncfa cookie (save this for API requests):")
            print(f"{ncfa_cookie}")

        return session, user

    except requests.exceptions.HTTPError as e:
        print(f"❌ HTTP Error: {e}")
        return None, None
    except requests.exceptions.RequestException as e:
        print(f"❌ Request Error: {e}")
        return None, None


def verify_sign_in(session):
    """Verify that sign in was successful."""
    try:
        response = session.get('https://www.geoguessr.com/api/v3/profiles')

        if response.ok:
            profile = response.json()
            print(f"\n✅ Verified: Signed in as {profile['user']['nick']}")
            return True
        else:
            print('❌ Not signed in')
            return False
    except Exception as e:
        print(f'Error verifying sign in: {e}')
        return False


def main():
    print("=" * 60)
    print("⚠️  WARNING: EDUCATIONAL USE ONLY")
    print("=" * 60)
    print("\nProgrammatic login may violate GeoGuessr's Terms of Service.")
    print("This example is for educational purposes to understand the API.")
    print("In production, use proper OAuth flows or manual cookie extraction.")
    print()

    proceed = input("Do you understand and wish to proceed? (yes/no): ").strip().lower()
    if proceed != 'yes':
        print("Exiting...")
        return

    # Get credentials from environment variables (NEVER hardcode!)
    email = os.getenv('GEOGUESSR_EMAIL')
    password = os.getenv('GEOGUESSR_PASSWORD')

    if not email or not password:
        print("\n❌ Error: Credentials not found in environment variables")
        print("\nSet them with:")
        print("  export GEOGUESSR_EMAIL='your_email@example.com'")
        print("  export GEOGUESSR_PASSWORD='your_password'")
        print("\n⚠️  Security Reminder:")
        print("- Never commit credentials to version control")
        print("- Use .env files with .gitignore")
        print("- Rotate passwords regularly")
        print("- Use strong, unique passwords")
        return

    # Attempt sign in
    session, user = sign_in(email, password)

    if session and user:
        # Verify sign in
        verify_sign_in(session)

        print("\n📝 Next Steps:")
        print("- Save your _ncfa cookie to GEOGUESSR_COOKIE environment variable")
        print("- Use that cookie for future API requests")
        print("- Don't share your cookie publicly")


if __name__ == '__main__':
    main()


"""
SECURITY BEST PRACTICES:

1. Environment Variables:
   - Store credentials in environment variables
   - Never hardcode in source code
   - Use .env files with python-dotenv

2. Password Security:
   - Use strong, unique passwords
   - Enable 2FA when available
   - Rotate credentials regularly

3. Session Management:
   - Store cookies securely
   - Implement timeout mechanisms
   - Clear sessions on logout

4. Rate Limiting:
   - Implement exponential backoff
   - Respect API rate limits
   - Monitor for unusual activity

5. Compliance:
   - Review Terms of Service
   - Obtain proper authorization
   - Document API usage
   - Respect user privacy

6. Error Handling:
   - Never log passwords
   - Handle authentication failures gracefully
   - Implement proper retry logic

7. Network Security:
   - Use HTTPS only
   - Verify SSL certificates
   - Implement timeout mechanisms
"""

File: examples/javascript/authentication/sign-in-example.js
/**
 * Sign In Example
 *
 * Demonstrates how to sign in to GeoGuessr programmatically.
 *
 * ⚠️  IMPORTANT WARNINGS:
 * 1. This is for EDUCATIONAL purposes only
 * 2. Programmatic login may violate GeoGuessr's Terms of Service
 * 3. May not work in browser due to CORS restrictions
 * 4. Use environment-specific authentication methods instead
 * 5. Never hardcode credentials in your code
 *
 * Usage in browser console (if allowed):
 * 1. Navigate to geoguessr.com
 * 2. Paste this code into the console
 * 3. Call: signIn('email@example.com', 'password')
 *
 * Note: This may only work in same-origin contexts or with proper CORS headers
 */

async function signIn(email, password) {
    try {
        const url = 'https://www.geoguessr.com/api/v3/accounts/signin';

        const response = await fetch(url, {
            method: 'POST',
            credentials: 'include',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                email: email,
                password: password
            })
        });

        if (!response.ok) {
            if (response.status === 401) {
                console.log('❌ Invalid email or password');
                return null;
            }
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const user = await response.json();

        // Display success
        console.log(`\n✅ Sign In Successful!\n`);
        console.log(`Welcome, ${user.nick}!`);
        console.log(`Level: ${user.progress.level}`);
        console.log(`XP: ${user.progress.xp.toLocaleString()}`);
        console.log(`Pro User: ${user.isProUser ? '✅ Yes' : '❌ No'}`);

        console.log(`\n🔐 Authentication Cookies Set:`);
        console.log(`- _ncfa (HttpOnly) - Main session cookie`);
        console.log(`- session - Session identifier`);
        console.log(`- devicetoken - Device tracking`);

        console.log(`\n💡 You can now make authenticated requests using credentials: 'include'`);

        return user;

    } catch (error) {
        console.error('❌ Sign in failed:', error.message);
        return null;
    }
}

// ⚠️  DO NOT USE THESE PATTERNS IN PRODUCTION:

// Bad example - hardcoded credentials:
// signIn('user@example.com', 'password123');

// Better example - use environment variables (Node.js):
// const email = process.env.GEOGUESSR_EMAIL;
// const password = process.env.GEOGUESSR_PASSWORD;
// signIn(email, password);

// Best practice - use proper OAuth flows or session management instead

/**
 * Check if sign in was successful
 */
async function verifySignIn() {
    try {
        const response = await fetch('https://www.geoguessr.com/api/v3/profiles', {
            credentials: 'include'
        });

        if (response.ok) {
            const profile = await response.json();
            console.log(`✅ Verified: Signed in as ${profile.user.nick}`);
            return true;
        } else {
            console.log('❌ Not signed in');
            return false;
        }
    } catch (error) {
        console.error('Error verifying sign in:', error);
        return false;
    }
}

// Example usage flow:
// async function authenticateAndVerify() {
//     const user = await signIn(email, password);
//     if (user) {
//         await verifySignIn();
//     }
// }

/**
 * Security Best Practices:
 *
 * 1. Never commit credentials to version control
 * 2. Use environment variables for sensitive data
 * 3. Implement rate limiting to prevent brute force
 * 4. Use HTTPS only
 * 5. Store passwords securely (hashed, salted)
 * 6. Implement proper session management
 * 7. Add 2FA when possible
 * 8. Rotate credentials regularly
 * 9. Monitor for suspicious login attempts
 * 10. Follow GeoGuessr's Terms of Service
 */
